README.rst

```diff
@@ -8,5 +8,195 @@
+  provides the underlying data storage, but accepting parameters, etc,
+  and marshaling them into the correct Python calls, is not a concern
+  here.
+- Providing a user interface or HTTPS API for viewing webhook audit
+  logs.
+
+In Scope/Features
+=================
+
+Certain concerns are very much in scope for this package, and this
+package should provide a complete, easy to use solution that addresses
+these concerns. Where necessary, if a concern cannot be addressed
+directly by this package, extension points (interfaces and
+``zope.component`` utilities) may be defined. These include, but are
+not limited to:
+
+- Resource Representation
+
+  The on-the-wire form of the resources is built using
+  `nti.externalization <https://ntiexternalization.readthedocs.io>`_.
+
+  To allow customization of the external forms, a named externalizer
+  is used; nti.externalization will fall back to the default
+  externalizer if no externalizer of the given name is available. The
+  default externalizer is named "webhook-delivery", but dialects may
+  use something different.
+
+- Alternate Webhook Dialects
+
+  Webhooks are a general protocol and mostly interoperable. But to
+  support cases where particular destinations have specific
+  requirements, "dialects" are used. There is a default dialect and
+  then there may be specializations of it. Each webhook subscription
+  may have associated with it the name of a dialect to use. These
+  dialects are found in the component registry. For example, a dialect
+  may choose to use a different externalizer name such as
+  "zapier-webhook-delivery".
+
+- Transactional
+
+  Webhooks should not be delivered if the ultimate creation or
+  persistence of a resource failed. To this end, webhook delivery in
+  this package is integrated with the `transaction
+  <https://transaction.readthedocs.io>`_ package.
+
+  Resources are externalized during a late phase of the transaction
+  commit process; the details about the delivery are recorded and
+  persisted, and only after the transaction is successfully committed
+  does the HTTP request get made.
+
+- Concurrency
+
+  Webhook delivery and record keeping should be lightweight, and
+  all actual network IO should proceed in a non-blocking fashion. This
+  means that this package will spawn threads (or greenlets, using
+  `gevent <http://www.gevent.org>`_.
+
+- Error Handling/Failure Retry
+
+  A limited amount of retry logic is provided by this package, but
+  that does not extend to process boundaries. If the process hosting
+  this package is killed while a delivery is pending, no automatic
+  provision is made to resume delivery attempts in any other process.
+
+  The API is present to allow that to be implemented, though.
+
+- Auditing/Delivery History
+
+  For each subscription, delivery attempts, status, and responses are
+  stored in a ring-buffer like structure. This can be inspected to see
+  if deliveries succeeded, failed, or never completed.
```

Comment:
Looks good to me. Stripe has done a really nice job with this. 

How much history do we expect to be able to keep? Is that a function of this package or consumers?

Do we keep failures for longer? indefinitely unless acknowledged?

Reply:
> How much history do we expect to be able to keep? Is that a function of this package or consumers?

Great question. I was expecting to define a default length here, and if-and-when transition to something configurable.

> Do we keep failures for longer? indefinitely unless acknowledged?
 
I wasn't intending to do anything special based on the status (keeps the ring buffer simple that way). I had arbitrarily been thinking to store the last 50 deliveries *per subscription*. If a subscription has 1 failure and then 51 successes, how important was that one failure? (I know, it depends. But for the use-cases we're talking about, if it was important enough to cause an issue (a) you know what was missed and (b) you can manually fire it.) 

I'm open to other positions, though I think "indefinitely" may not be a reasonable one.